zc_utils.py

```diff
@@ -10,18 +13,58 @@ def write_compact_size(n):
     else:
         return struct.pack('B', 255) + struct.pack('<Q', n)
 
-assert write_compact_size(0) == b'\x00'
-assert write_compact_size(1) == b'\x01'
-assert write_compact_size(252) == b'\xFC'
-assert write_compact_size(253) == b'\xFD\xFD\x00'
-assert write_compact_size(254) == b'\xFD\xFE\x00'
-assert write_compact_size(255) == b'\xFD\xFF\x00'
-assert write_compact_size(256) == b'\xFD\x00\x01'
-assert write_compact_size(0xFFFE) == b'\xFD\xFE\xFF'
-assert write_compact_size(0xFFFF) == b'\xFD\xFF\xFF'
-assert write_compact_size(0x010000) == b'\xFE\x00\x00\x01\x00'
-assert write_compact_size(0x010001) == b'\xFE\x01\x00\x01\x00'
-assert write_compact_size(0xFFFFFFFE) == b'\xFE\xFE\xFF\xFF\xFF'
-assert write_compact_size(0xFFFFFFFF) == b'\xFE\xFF\xFF\xFF\xFF'
-assert write_compact_size(0x0100000000) == b'\xFF\x00\x00\x00\x00\x01\x00\x00\x00'
-assert write_compact_size(0xFFFFFFFFFFFFFFFF) == b'\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF'
+def parse_compact_size(rest):
+    assert len(rest) >= 1
+    b = rest[0]
+    if b < 253:
+        return (b, rest[1:])
+    elif b == 253:
+        assert len(rest) >= 3
+        n = struct.unpack('<H', rest[1:3])[0]
+        assert n >= 253
+        return (n, rest[3:])
+    elif b == 254:
+        assert len(rest) >= 5
+        n = struct.unpack('<I', rest[1:5])[0]
+        assert n >= 0x10000
+        return (n, rest[5:])
+    else:
+        assert len(rest) >= 9
+        n = struct.unpack('<Q', rest[1:9])[0]
+        assert n >= 0x100000000
+        return (n, rest[9:])
```

Comment:
The parser should also (perhaps optionally) enforce the max size limit (so at the parse callsite we can enforce the intended behaviour).